ain/utils/V3Keystore.py

```diff
@@ -0,0 +1,294 @@
+        self.address = address
+        self.crypto = Crypto()
+        self.crypto.ciphertext = ciphertext
+        self.crypto.cipherparams = Cipherparams()
+        self.crypto.cipherparams.iv = iv
+        self.crypto.cipher = cipher
+        self.crypto.kdf = kdf
+        self.crypto.kdfparams = kdfparams
+        self.crypto.mac = mac
+
+    @classmethod
+    def fromPrivateKey(
+        cls,
+        privateKey: Union[bytes, str],
+        password: str,
+        options: V3KeystoreOptions = V3KeystoreOptions()
+    ):
+        """
+        Converts an account into a V3 Keystore and encrypts it with a password.
+        """
+
+        if type(privateKey) is str:
+            privateKeyBytes = bytes.fromhex(privateKey)
+        elif type(privateKey) is bytes:
+            privateKeyBytes = privateKey
+        else:
+            raise ValueError("Invalid public key type")
+
+        salt = getattr(options, "salt", token_bytes(32))
+        iv = getattr(options, "iv", token_bytes(16))
+        uuid = getattr(options, "uuid", token_bytes(16))
+        dklen = getattr(options, "dklen", 32)
+        kdf = getattr(options, "kdf", "scrypt")
+
+        kdfparams = {
+            "dklen": dklen,
+            "salt": salt.hex(),
+        }
+        if kdf == "scrypt":
+            n = getattr(options, "n", 262144)
+            r = getattr(options, "r", 8)
+            p = getattr(options, "p", 1)
+            kdfparams.update({
+                "n": n,
+                "r": r,
+                "p": p,
+            })
+            derivedKeyTemp = scrypt(
+                password,
+                salt, # type: ignore
+                dklen,
+                N=n,
+                r=r,
+                p=p
+            )
+            if type(derivedKeyTemp) is bytes:
+                derivedKey = derivedKeyTemp
+            else:
+                raise ValueError("scrypt derived multiple keys")
+        elif kdf == "pbkdf2":
+            c = getattr(options, "c", 1000)
+            kdfparams.update({
+                "c": c,
+                "prf": "hmac-sha256",
+            })
+            derivedKey = PBKDF2(
+                password,
+                salt,
+                dklen,
+                count=c,
+                hmac_hash_module=SHA256
+            )
+        else:
+            raise ValueError("Unsupported kdf")
+
+        cipherkey = derivedKey[0:16]
+        padding = False
+        cipher = getattr(options, "cipher", "aes-128-ctr")
+        if cipher == "aes-128-ecb":
+            aescipher = AES.new(cipherkey, AES.MODE_ECB)
+            padding = True
+        elif cipher == "aes-128-cbc" or cipher == "aes128":
+            aescipher = AES.new(cipherkey, AES.MODE_CBC, iv)
+            padding = True
+        elif cipher == "aes-128-cfb":
+            aescipher = AES.new(cipherkey, AES.MODE_CFB, iv, segment_size=128)
+        elif cipher == "aes-128-cfb8":
+            aescipher = AES.new(cipherkey, AES.MODE_CFB, iv)
```

Comment:
segment_size=8 probably?

Reply:
Oh, never mind. default value is 8!

Reply:
`aes-128-cfb8` was that, and this agreed with ain-js.

Reply:
I think this link helps. [Link](https://crypto.stackexchange.com/questions/79654/what-is-the-difference-between-cfb-and-cfb1-and-cfb8-like-openssls-aria-128)

Reply:
Huh, your comment didn't update well, sorry.